### PR TITLE
Add tests for the compat macro

### PR DESCRIPTION
**Pull request by Elchi3**
We need to write tests for the compat macro. The complexity of the compat data is too damn high! We're discovering more and more (small) table rendering bugs, which we fix, but we have no way to tell if regress something else. Soon, we want rewrite the rendering for a redesign and I would like to avoid further regressions.

This is an unfinished PR. I've never written a test for a KS macro before and I need help with mocha, sinon, and so forth. The good news is: I managed to add a very simple test!

Now I have no idea how to stub the mdn-browser-compat-data package. I added "data" and "expected" fixtures, but I'm unclear how to get the compat.ejs macro make use of this. Can I get some hints, @escattone?

### CODE REVIEW

**Comment by escattone** [comment]
Thanks @Elchi3, this looks like a great start. I'll take a look.

File: tests/macros/test-compat.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    );
+
+    fixtures[name].expected =
+        fs.readFileSync(path.resolve(fixture_dir, fixtures[name].expectedFile), 'utf8');
+}
+
+describeMacro('Compat', function () {
+
+    beforeEachMacro(function (macro) {
+        // insert magic here to make bcd fixtures available
```
Comment (escattone) [inline tests/macros/test-compat.js]: Here's what I would add here:
```
macro.ctx.require = sinon.stub();
macro.ctx.require.withArgs('mdn-browser-compat-data').returns(
    fixtures.alternative_name.data
);
```
But when I do that, it seems like the call to `require('mdn-browser-compat-data')` (which should call the stub) is not returning the fixture data. I need to look into that more on Monday.

**Comment by escattone** [comment]
Sorry for the delay @Elchi3 ! I'm going to try to address this in this calm period of a day or two before we "go-live" with AWS.

**Comment by Elchi3** [comment]
No worries @escattone! This is not that important right now :)

File: tests/macros/fixtures/compat/alternative_name.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{
+    "webextensions": {
+        "api:": {
```
Comment (escattone) [inline tests/macros/fixtures/compat/alternative_name.json]: This typo was difficult for me to recognize! Should be `"api"` not `"api:"`

**Comment by escattone** [comment]
@Elchi3 If you fix the typo noted above, and make your `beforeEachMacro` look like this:
```
beforeEachMacro(function (macro) {
    macro.ctx.require = sinon.stub();
    macro.ctx.require.withArgs('mdn-browser-compat-data').returns(
        fixtures.alternative_name.data
    );
});
```
That should get you on your way. It looks like the expected data (`kumascript/tests/macros/fixtures/compat/alternative_name-expected.txt`) might have to be modified, as the second test fails, but of course that's for you to decide!

**Comment by escattone** [comment]
By the way, thanks @Elchi3 for adding these tests!

File: tests/macros/fixtures/compat/alternative_name.json (hunk 2 above)
Comment (Elchi3) [reply tests/macros/fixtures/compat/alternative_name.json]: woah, sorry about that!

**Comment by Elchi3** [comment]
Thanks for looking into this! It works well (sorry again for the frustrating typo). Now I just have to see about white space and \n when doing the comparison. Will do more here once I'm done with the current sprint.

**Comment by Elchi3** [comment]
I think I'm not happy with this approach. As you can see, the assertion fails as the comparison of the two HTML strings (one from {{compat}}, one from a file containing the expected HTML) doesn't work due to different new lines and and whitespace etc.

I think asserting against a DOM would be easier. Basically, almost all KumaScript macros generate HTML snippets (which always will be a fragment DOM). For some, this DOM is smaller (like for the existing tests, see e.g. test-draft.js), but for some it is also larger (this one, sidebars, more complex banners, etc).

I don't know, but maybe something like jsdom could helps us as a dev dependency in our tests, see e.g. https://github.com/tmpvar/jsdom#fragment

```
const compat = JSDOM.fragment(compatTemplateOutput);
assertEqual(compat.childNodes.length, 2);
assertEqual(compat.querySelector("p").textContent, "Yup, this is the compat note's text");
```

**Comment by Elchi3** [comment]
I just added a commit that uses jsdom. I think I like that a lot more and, as said above, we're always facing HTML snippets from macros, so this could be really useful when writing KS tests generally. Opinions?

**Pull request merged**